dep/wire.go

```diff
@@ -68,11 +69,22 @@ var observabilitySet = wire.NewSet(
 	mdtracer.NewLocal,
 )
 
+func makeListenerList(
```

Comment:
And no logic here.

Reply:
`wire` has difficulties with multiple providers configuration, please take a look [here](https://github.com/google/wire/issues/207). So it is a way how we can do it.

Do you have an idea how to do it?

Reply:
Any thoughts about this? https://github.com/short-d/short/blob/master/backend/app/adapter/routing/route.go#L35-L56

For your information, I am moving the routes into use case.

```go
type UseCase interface {
	ViewLongLink(
		alias string,
		aliasNotFoundCallback func(),
		viewLongLinkCallback func(longLink string),
	)
}

var _ UseCase = (*Short)(nil)

type Short struct {
	logger       fw.Logger
	urlRetriever url.Retriever
	timer        fw.Timer
}

func (s Short) ViewLongLink(
	alias string,
	aliasNotFoundCallback func(),
	viewLongLinkCallback func(longLink string),
) {
	now := s.timer.Now()
	u, err := s.urlRetriever.GetURL(alias, &now)
	if err != nil {
		s.logger.Error(err)
		aliasNotFoundCallback()
		return
	}
	viewLongLinkCallback(u.OriginalURL)
}

func NewShort(
	logger fw.Logger,
	urlRetriever url.Retriever,
	timer fw.Timer,
) Short {
	return Short{
		logger:       logger,
		urlRetriever: urlRetriever,
		timer:        timer,
	}
}
```

```go
// NewOriginalURL translates alias to the original long link.
func NewOriginalURL(
	tracer fw.Tracer,
	useCase usecase.UseCase,
	webFrontendURL netURL.URL,
) fw.Handle {
	return func(w http.ResponseWriter, r *http.Request, params fw.Params) {
		trace := tracer.BeginTrace("OriginalURL")
		alias := params["alias"]

		useCase.ViewLongLink(alias, func() {
			serve404(w, r, webFrontendURL)
			trace.End()
		}, func(longLink string) {
			http.Redirect(w, r, longLink, http.StatusSeeOther)
			trace.End()
		})
	}
}
```

```
// NewShortRoutes creates HTTP routes for Short API with WwwRoot to uniquely identify WwwRoot during dependency injection.
func NewShortRoutes(
	logger fw.Logger,
	tracer fw.Tracer,
	webFrontendURL WebFrontendURL,
	useCase usecase.UseCase,
	githubAPI github.API,
	facebookAPI facebook.API,
	googleAPI google.API,
	authenticator auth.Authenticator,
	accountProvider account.Provider,
) []fw.Route {
	observability := routing.Observability{
		Logger: logger,
		Tracer: tracer,
	}

	return routing.NewShort(
		observability,
		string(webFrontendURL),
		useCase,
		githubAPI,
		facebookAPI,
		googleAPI,
		authenticator,
		accountProvider,
	)
}
```